File: src/143_Reorder_List.py
# 143. Reorder List
# https://leetcode.cn/problems/reorder-list/

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
            
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        slow, fast = head, head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        
        right = slow.next
        slow.next = None # 这一步别忘记了
        right = self.reverse_list(right)
        
        cur = head
        while right and cur:
            cur_sec = right
            right = right.next
            next_cur = cur.next
            cur_sec.next = cur.next
            cur.next = cur_sec
            cur = next_cur
        
    def reverse_list(self, head):
        previous = None
        current = head
        while current is not None:
            next_node = current.next
            current.next = previous
            previous = current
            current = next_node
        return previous

'''
反转错了
    1.快慢指针找到中点 
    2.拆成两个链表 
    3.遍历两个链表，后面的塞到前面的“缝隙里”
'''
# class Solution:
#     def reorderList(self, head: Optional[ListNode]) -> None:
#         """
#         Do not return anything, modify head in-place instead.
#         """
#         slow, fast = head, head
#         while fast and fast.next:
#             slow = slow.next
#             fast = fast.next.next
        
#         right = slow.next
#         slow.next = None
#         slow = self.reverse_list(head)

#         left_curr = slow.next
#         left_prev = slow
#         right_curr = right.next
#         right_prev = right
#         while right_prev:
#             left_prev.next = right_prev
#             left_prev = left_curr
#             left_curr = left_curr.next
#             right_prev.next = left_prev
#             right_prev = right_curr
#             if right_curr:
#                 right_curr = right_curr.next
        
#         self.reverse_list(slow)
        
#     def reverse_list(self, head):
#         previous = None
#         current = head
#         while current is not None:
#             next_node = current.next
#             current.next = previous
#             previous = current
#             current = next_node
#         return previous
        
        
# C++ , 不得不承认这个题挺好的，考察了很多链表的基本功
# 整体思路：
# 1、快慢指针找到中间节点
# 2、反转后半段链表
# 3、合并前半段链表和后半段链表
class Solution:
    def reorderList(self, head: ListNode) -> None:
        if not head:
            return
        
        mid = self.middleNode(head)
        l1 = head
        l2 = mid.next
        mid.next = None
        l2 = self.reverseList(l2)
        self.mergeList(l1, l2)
    
    def middleNode(self, head: ListNode) -> ListNode:
        slow = fast = head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
        return slow
    
    def reverseList(self, head: ListNode) -> ListNode:
        prev = None
        curr = head
        while curr:
            nextTemp = curr.next
            curr.next = prev
            prev = curr
            curr = nextTemp
        return prev

    def mergeList(self, l1: ListNode, l2: ListNode):
        while l1 and l2:
            l1_tmp = l1.next
            l2_tmp = l2.next

            l1.next = l2
            l1 = l1_tmp

            l2.next = l1
            l2 = l2_tmp

File: src/143_Reorder_List.java
// 143_Reorder_List
// https://leetcode.cn/problems/reorder-list/


// Definition for singly-linked list.
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}
 

// class Solution {
//     public void reorderList(ListNode head) {
//         if (head == null || head.next == null || head.next.next == null) {
//             return;
//         }
        
//         ListNode fast = head;
//         ListNode slow = head;

//         while (fast != null && fast.next != null) {
//             fast = fast.next.next;
//             slow = slow.next;
//         }

//         ListNode head2 = slow.next;
//         slow.next = null;

//         ListNode rev_head2 = reverseList(head2);

//         System.out.println(rev_head2);

//         ListNode c1 = head;
//         ListNode n1 = head.next;
//         ListNode c2 = rev_head2;
//         ListNode n2 = rev_head2.next;

//         while (n1 != null && c2 != null ) {
//             c1.next = c2;
//             c2.next = n1;
//             c1 = n1;
//             if (n1 != null) {
//                 n1 = n1.next;
//             }
//             c2 = n2;
//             if (n2 != null) {
//                 n2 = n2.next;
//             }
//         }

//         return;
//     }

//     public ListNode reverseList(ListNode head) {
//         if (head == null || head.next == null) {
//             return head;
//         }
//         ListNode pre = null;
//         ListNode cur = head;
//         ListNode nex = head.next;

//         while (cur != null) {
//             cur.next = pre;
//             pre = cur;
//             cur = nex;
//             if (nex != null) {
//                 nex = nex.next;
//             }
//         }

//         return pre;
//     }
// }


class Solution {
    public void reorderList(ListNode head) {
        if (head == null || head.next == null || head.next.next == null) {
            return;
        }

        ListNode fast = head, slow = head;
        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;
        }

        ListNode head2 = slow.next;
        slow.next = null;
        ListNode revHead2 = reverseList(head2);

        ListNode c1 = head, c2 = revHead2;
        while (c1 != null && c2 != null) {
            ListNode n1 = c1.next, n2 = c2.next;
            c1.next = c2;
            c2.next = n1;
            c1 = n1;
            c2 = n2;
        }
    }

    public ListNode reverseList(ListNode head) {
        ListNode pre = null, cur = head;
        while (cur != null) {
            ListNode nex = cur.next;
            cur.next = pre;
            pre = cur;
            cur = nex;
        }
        return pre;
    }
}
